File: question/neetcode150/binary_search/time-based-key-value-store.py
from collections import defaultdict

class TimeMap:

    def __init__(self):
        self.d = defaultdict(list)
        

    def set(self, k: str, v: str, t: int) -> None:
        self.d[k].append((v,t))
        

    def get(self, k: str, t: int) -> str:
        a = self.d[k]

        i = 0
        n = len(a)
        while n:
            while i + n < len(a) and a[i+n][1] <= t:
                i+=n
            n//=2

        return a[i][0] if a[i][1] <= t else ""


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)

File: question/neetcode150/binary_search/time-based-key-value-store.cpp
#include <bits/stdc++.h>

using namespace std;

class TimeMap {
public:
    unordered_map<string, vector<pair<int, string>>> d;

    TimeMap() {
    }
    
    void set(string k, string v, int t) {
        d[k].push_back({t,v});
    }
    
    string get(string s, int t) {
        int n = d[s].size();

        int k = -1;
        for(int i=n;i;i/=2){
            while (k+i<n and d[s][k+i].first<=t){
                k+=i;
            }
        }

        if(k==-1) return "";
        return d[s][k].second;
    }
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */